### ISSUE TITLE
Optimizing for delay vs. quality?

### ISSUE THREAD

**Issue by Tryptophan** [t-mullen/SparrowTV#3]
To my knowledge WebRTC will prioritize keeping a low delay between two participants in a call and will automatically adjust resolution and bitrate to make up for loss in connection speed. Is there a way to do the opposite and prioritize the quality of the stream over the delay time? I would assume for SparrowTV it would be better to prioritize a high quality stream over a low delay since it's one to many.

**Comment by RationalCoding** [comment]
You can set minimum bitrate (and quality) through the SDP.

https://webrtchacks.com/limit-webrtc-bandwidth-sdp/

**Comment by Tryptophan** [comment]
That is exactly what I was looking for! Thanks!

**Comment by Tryptophan** [comment]
Do you know if there is a way to increase the buffer size of the stream being sent through the SDP? I'm not sure setting a minimum bandwidth or quality would achieve what I want, as I'm not 100% sure that will automatically increase the buffer size.

**Comment by Tryptophan** [comment]
I've found this: https://stackoverflow.com/questions/33130262/is-buffering-possible-in-webrtc. But I'm not sure if that is just for audio. (Also a lot of the codec related stuff is way over my head)

**Comment by t-mullen** [comment]
So basically you want to modify the congestion control to buffer rather than lose quality when the network drops packets?

ie) bad network leads to “buffering” rather than quality drops, like in traditional video streaming.

This is a bit odd, since one of the big features of WebRTC is this adaptive quality approach to congestion control... but I can definitely see the use case. 

From what I can tell, the only way you might be able to adjust this is through codec-specific options in the SDP: like minptime in Opus https://tools.ietf.org/html/draft-spittka-payload-rtp-opus-03#page-12

Basically, force a high quality at the codec level and WebRTC will handle buffers for you.

**Comment by Tryptophan** [comment]
The reason I want a larger buffer is that over multiple connections between chained peers, even with a good upload speed, there is a drop in quality due to WebRTC trying to reduce delay. Even over just 2 or 3 chained connections the resolution on video streams will drop from 1080p to 480p. I would assume you would want the same for SparrowTV since it's traditional video streaming as well.

**Comment by t-mullen** [comment]
Right, the lossy connections are a big problem when you're making long chains of peers.

You might want to consider using the datachannel to buffer/send video data. This would be as lossless as you want, and you could even switch to lossy/performant stream in the final layer of the stream. You'll have to do this in any real application in order to validate data anyhow. WebTorrent uses this approach to send video data and seems to have no performance issues.

**Comment by Tryptophan** [comment]
Does SparrowTV currently use the datachannel? Asking cause it would be good to get an example that I can look at. Thank you for all the help btw, this stuff is super hard to find just looking through documentation/google.

**Comment by t-mullen** [comment]
No, the demo uses MediaStreams. The underlying module uses [peer-tree](https://github.com/t-mullen/peer-tree), which supports both. 

👍

**Comment by t-mullen** [comment]
Are you chunking your data? There's a maximum of ~16kb for a call to `send`. The sending peer should throw an error event if this is happening though.

ie)
```javascript
var chunks = require('chunk-stream')
inputStream.pipe(chunks(16000)).pipe(peer)
```

**Comment by Tryptophan** [comment]
I have done that, and the data seems to get across fine now, but there is an issue when reading the stream in from multiple peers. I've noticed one-to-one streaming using MediaStream=>MediaRecorder=>MediaSource seems to playback fine. But when I try to append chunks using the same data with another peer (2 or more), I'm getting a "CHUNK_DEMUXER_ERROR_APPEND_FAILED" error. I assume this is because I can't just start appending webm chunks (vp9/opus), but this is all I could find on the issue: https://stackoverflow.com/questions/50221344/mediasource-buffer-works-on-only-one-client.

**Comment by t-mullen** [comment]
Yes, you need the cluster header and full clusters. I've already made a module for this:
https://github.com/t-mullen/mediastream-to-webm

You can see it in action here: https://github.com/t-mullen/wbs-plus-hypercast (This uses Beaker's Dat protocol instead of WebRTC datachannels, but the concept is the same.)
